### PR TITLE
Add solutions to exercises

### PR DESCRIPTION
**Pull request by jdenholm**
Adding full worked solutions to each exercise:

- [x] Penguin classification.
- [x] Penguin regression.
- [x] MNIST classification.
- [x] CNN regression.

### CODE REVIEW

**Review by jatkinson1000 — APPROVED** [review]
I've been through all of these as individual PRs, and again here (see a few more typo fixes).

I'm now happy that they match to the exercises, are correct and functional, and make sense.

However, it would be good to have an independent pair of eyes look over them so I'll leave this open for now in case anyone else has a chance to look.

**Review by mondus — CHANGES_REQUESTED** [review]
I ran out of time today to go through this in details but checked the first two examples worked in the notebook by running them through. Can confirm they work as expected. I would however suggest/request to remove all cell outputs before committing so that users have a "clean" workbook prior to running cells.

**Review by TomMelt — COMMENTED** [review]
I have a couple of suggestions and I have also submitted a PR #23  with some typos fixes. I recommend merging that and then you can optionally address the comments below.

**Comment by jatkinson1000** [comment]
@mondus We cleared the books for the lab, but since these are worked solutions I feel there is some merit to having example output present should people just be reading and not running, or running their version in parallel.
Any thoughts @jdenholm ?

File: exercise-solutions/02_penguin_regression_solutions.ipynb [hunk 1]
```diff
@@ -0,0 +1,813 @@
...
+    "axis.set_ylabel(\"Loss (MSE)\", fontsize=15)\n",
+    "\n",
+    "fig.tight_layout()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "While this losses have decreased nicely, MSE isn't the most easily interpretable way to evaluate the model's performance. We could instead look at the ratio of the predictions  targets—and in particular, the distribution of this quantity."
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply exercise-solutions/02_penguin_regression_solutions.ipynb:741]: Should be a 'to' in there, I'll fix.

File: exercise-solutions/02_penguin_regression_solutions.ipynb [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+    "loss_func = MSELoss()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Task 8: Selecting an optimiser\n",
+    "\n",
+    "``Adam`` is regarded as the king of optimisers: let's use it again, but this time with a learning rate.\n",
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply exercise-solutions/02_penguin_regression_solutions.ipynb:463]: This was done in the slides, but worth mentioning again/referring to lectures. I'll add.

File: exercise-solutions/01_penguin_classification_solutions.ipynb [hunk 3]
```diff
@@ -0,0 +1,897 @@
...
+    "    Parameters\n",
+    "    ----------\n",
+    "    in_feats : int\n",
+    "        The number of input features to the model.\n",
+    "    out_feats : int\n",
+    "        The number of output features (or classes) the model should produce.\n",
+    "\n",
+    "    Notes\n",
+    "    -----\n",
+    "    The ``torch.nn.Sequential`` allows us to \"chain\" multiple layers, rather\n",
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply exercise-solutions/01_penguin_classification_solutions.ipynb:425]: Will amend.

File: exercise-solutions/02_penguin_regression_solutions.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply exercise-solutions/02_penguin_regression_solutions.ipynb:463]: Link to Adam Docs is in there.

File: exercise-solutions/01_penguin_classification_solutions.ipynb [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+   "source": [
+    "from torch.utils.data import DataLoader\n",
+    "\n",
+    "batch_size = 16\n",
+    "\n",
+    "train_loader = DataLoader(\n",
+    "    train_set,\n",
+    "    batch_size=batch_size,\n",
+    "    shuffle=True,\n",
+    "    drop_last=True,\n",
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply exercise-solutions/01_penguin_classification_solutions.ipynb:358]: Added comment.

**Review by TomMelt — APPROVED** [review]

**Comment by jdenholm** [comment]
> @mondus We cleared the books for the lab, but since these are worked solutions I feel there is some merit to having example output present should people just be reading and not running, or running their version in parallel. Any thoughts @jdenholm ?

I can see your point as much as I can see Paul's. If the user doesn't have a particularly beefy laptop, they might find running the CNN examples challenging, so I think I lean slightly towards leaving them in.

**Comment by jatkinson1000** [comment]
Cool, thanks.
In that case if you're happy to approve this for merging @mondus I'll get it done and sent to participants.

**Comment by jatkinson1000** [comment]
This will close #13

**Review by mondus — APPROVED** [review]

**Pull request merged**